### PR TITLE
Lora support

### PR DESCRIPTION
**Pull request by fabik111**
Support for LoRaWAN boards.
ArduinoIoTCloud has been separated in ArduinoIoTCloudTCP and ArduinoIoTCloudLPWAN handlers to support TCP and Lora connection. ArduinoIoTCloud becomes a pure virtual class. 
One of these two object is instantiated according board type

### CODE REVIEW

File: src/ArduinoIoTCloud.h [hunk 1]
```diff
@@ -145,114 +115,145 @@ class ArduinoIoTCloudClass {
...
-      return _brokerPort;
+
+    ArduinoCloudProperty& addPropertyReal(String& property, String name, int tag, Permission const permission) {
+      ArduinoCloudProperty *p = new CloudWrapperString(property);
+      return Thing.addPropertyReal(*p, name, permission, tag);
     }
-    void printDebugInfo();
-    void addCallback(ArduinoIoTCloudEvent const event, OnCloudEventCallback callback);
+
+    void addCallback(ArduinoIoTCloudEvent const event, OnCloudEventCallback callback) {
```
[the remark this answers is not in the document]
Comment (fabik111) [reply src/ArduinoIoTCloud.h]: fixed

File: src/ArduinoIoTCloud.h [hunk 2]
```diff
@@ -145,114 +115,145 @@ class ArduinoIoTCloudClass {
...
-
-    static void execCloudEventCallback(OnCloudEventCallback & callback, void * callback_arg);
-    static void printConnectionStatus(ArduinoIoTConnectionStatus status);
+    int _lastSyncRequestTickTime = 0;
+
+    OnCloudEventCallback _on_sync_event_callback = NULL;
+    OnCloudEventCallback _on_connect_event_callback = NULL;
+    OnCloudEventCallback _on_disconnect_event_callback = NULL;
+
+    static void execCloudEventCallback(OnCloudEventCallback & callback, void * callback_arg) {
```
[the remark this answers is not in the document]
Comment (fabik111) [reply src/ArduinoIoTCloud.h]: fixed

File: src/ArduinoIoTCloud.h [hunk 3]
```diff
@@ -145,114 +115,145 @@ class ArduinoIoTCloudClass {
...
+      if (callback) {
+        (*callback)(callback_arg);
+      }
+    }
+    static void printConnectionStatus(ArduinoIoTConnectionStatus status) {
```
[the remark this answers is not in the document]
Comment (fabik111) [reply src/ArduinoIoTCloud.h]: fixed

File: src/ArduinoIoTCloud.h [hunk 4]
```diff
@@ -78,63 +63,48 @@ enum class ArduinoIoTCloudEvent {
...
-    inline ConnectionHandler * getConnection() {
-      return _connection;
-    }
 
 #define addProperty( v, ...) addPropertyReal(v, #v, __VA_ARGS__)
 
     static unsigned long const DEFAULT_MIN_TIME_BETWEEN_UPDATES_MILLIS = 500; /* Data rate throttled to 2 Hz */
 
-    void addPropertyReal(ArduinoCloudProperty& property, String name, permissionType permission_type = READWRITE, long seconds = ON_CHANGE, void(*fn)(void) = NULL, float minDelta = 0.0f, void(*synFn)(ArduinoCloudProperty& property) = CLOUD_WINS) {
+    void addPropertyReal(ArduinoCloudProperty & property, String name, permissionType permission_type = READWRITE, long seconds = ON_CHANGE, void(*fn)(void) = NULL, float minDelta = 0.0f, void(*synFn)(ArduinoCloudProperty & property) = CLOUD_WINS) {
```
[the remark this answers is not in the document]
Comment (fabik111) [reply src/ArduinoIoTCloud.h]: fixed

File: src/ArduinoIoTCloudLPWAN.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  public:
+    ArduinoIoTCloudLPWAN();
+    ~ArduinoIoTCloudLPWAN();
+    int connect();
+    bool disconnect();
+    int connected();
+    inline void update() {
+      update(NULL);
+    }
+    inline void update(int const reconnectionMaxRetries, int const reconnectionTimeoutMs) __attribute__((deprecated)) {
```
[the remark this answers is not in the document]
Comment (fabik111) [reply src/ArduinoIoTCloudLPWAN.h]: I can't move `update()` and `update(int const reconnectionMaxRetries, int const reconnectionTimeoutMs)`  to `ArduinoIoTCloud ` because both call `update(CallbackFunc onSyncCompleteCallback)` that must be virtual... I've tried this solution:
```
ArduinoIoTCloud.h
inline void update() {
      update(NULL);
    }
    inline void update(int const reconnectionMaxRetries, int const reconnectionTimeoutMs) __attribute__((deprecated)) {
      update(NULL);
    }
virtual void update(CallbackFunc onSyncCompleteCallback) __attribute__((deprecated)) = 0;
```
but compiler can't resolve correctly the calling `ArduinoCloud.update()` unless I change the firm from `virtual void update(CallbackFunc onSyncCompleteCallback) __attribute__((deprecated)) = 0;`  to `virtual void update1(CallbackFunc onSyncCompleteCallback) __attribute__((deprecated)) = 0;` but for backward compatibility I can't

File: src/ArduinoIoTCloudTCP.cpp [hunk 6]
```diff
@@ -29,21 +29,21 @@
 #endif
 
 #ifdef BOARD_HAS_ECCX08
-  const static int keySlot                                   = 0;
-  const static int compressedCertSlot                        = 10;
+  const static int keySlot = 0;
```
[the remark this answers is not in the document]
Comment (fabik111) [reply src/ArduinoIoTCloudTCP.cpp]: fixed

File: src/ArduinoIoTCloudLPWAN.h [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+static uint8_t const DEFAULT_CBOR_LORA_MSG_SIZE = 255;
+
+class ArduinoIoTCloudLPWAN : public ArduinoIoTCloudClass {
+  public:
+    ArduinoIoTCloudLPWAN();
+    ~ArduinoIoTCloudLPWAN();
+    int connect();
+    bool disconnect();
+    int connected();
+    inline void update() {
```
[the remark this answers is not in the document]
Comment (fabik111) [reply src/ArduinoIoTCloudLPWAN.h]: `ArduinoIoTCloud` should be an interface for future implementation of other protocols, in this case they are almost the same but tomorrow could not

File: src/ArduinoIoTCloud.cpp [hunk 8]
```diff
@@ -17,444 +17,117 @@
...
+void ArduinoIoTCloudClass::execCloudEventCallback(OnCloudEventCallback& callback, void* callback_arg) {
   if (callback) {
     (*callback)(callback_arg);
   }
 }
-
 void ArduinoIoTCloudClass::printConnectionStatus(ArduinoIoTConnectionStatus status) {
   switch (status) {
-    case ArduinoIoTConnectionStatus::IDLE:         Debug.print(DBG_INFO,  "Arduino IoT Cloud Connection status: IDLE");         break;
+    case ArduinoIoTConnectionStatus::IDLE:         Debug.print(DBG_INFO, "Arduino IoT Cloud Connection status: IDLE");         break;
```
Comment (lxrobotics) [inline src/ArduinoIoTCloud.cpp]: This is a unnecessary formatting change which enlarges the change set that needs to be reviewed without any functional change. Please refrain from formatting changes when doing functional changes and restore the original version.

File: src/ArduinoIoTCloudLPWAN.cpp [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+ArduinoIoTCloudLPWAN::ArduinoIoTCloudLPWAN() :
+  _connection(NULL) {}
+
+ArduinoIoTCloudLPWAN::~ArduinoIoTCloudLPWAN() {
+}
+
+int ArduinoIoTCloudLPWAN::connect() {
+  _connection->connect();
+  int state = _connection->getStatus() == NetworkConnectionState::INIT ? 1 : 0;
```
Comment (lxrobotics) [inline src/ArduinoIoTCloudLPWAN.cpp]: Minor improvement suggestion: Since state is written-once and read-only after you can make it `const`.

File: src/ArduinoIoTCloudLPWAN.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (lxrobotics) [reply src/ArduinoIoTCloudLPWAN.h]: I have to disagree here because
`inline void update(int const reconnectionMaxRetries, int const reconnectionTimeoutMs) __attribute__((deprecated))`
and
`void update(CallbackFunc onSyncCompleteCallback) __attribute__((deprecated))`
will be removed soon since they have been marked `__attribute__((deprecated))` for some time. What remains would be a pure virtual `update` in `ArduinoIoTCloud` which is then implemented in the derived base classes. However, let's not do this change now.

File: src/ArduinoIoTCloud.cpp (hunk 8 above)
Comment (fabik111) [reply src/ArduinoIoTCloud.cpp]: fixed

File: src/ArduinoIoTCloudLPWAN.cpp (hunk 9 above)
Comment (fabik111) [reply src/ArduinoIoTCloudLPWAN.cpp]: fixed

File: src/CloudSerial.cpp [hunk 10]
```diff
@@ -14,6 +14,7 @@
    software without disclosing the source code of your own applications. To purchase
    a commercial license, send an email to license@arduino.cc.
 */
+#if defined(ARDUINO_SAMD_MKRGSM1400) || defined(ARDUINO_SAMD_MKRWIFI1010) ||  defined(ARDUINO_SAMD_MKR1000) || defined(ARDUINO_SAMD_NANO_33_IOT) || defined(ARDUINO_ESP8266_ESP12) || defined(ARDUINO_ARCH_ESP32) || defined(ESP8266)
```
Comment (AlbyIanna) [inline src/CloudSerial.cpp]: @fabik111 Why is `ARDUINO_SAMD_MKRNB1500` not here?

Wouldn't it be better if we use an `ifndef` instead?
```suggestion
#ifndef HAS_LORA
```
Like this, it wouldn't be necessary to edit this line every time we add a new board (and it's less likely that we forget to add some)

Comment (fabik111) [reply src/CloudSerial.cpp]: Fixed, pay attention to the future boards that will not support cloudSerial
